File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    unordered_map<int,Node*>map;
    Node* cloneGraph(Node* node) {
        if(node==NULL) return node;
        if(map.find(node->val)!=map.end()){
            return map[node->val];
        }
        Node* copy = new Node(node->val, vector<Node*>());
        map[node->val]=copy;
        for(Node* neigh:node->neighbors){
            copy->neighbors.push_back(cloneGraph(neigh));
        }
        return copy;
    }
};

File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if(node==null) return null;
        HashMap<Node,Node>map = new HashMap<>();
        return dfs(node, map);
    }
    public Node dfs(Node head, HashMap<Node,Node>map){
        if(head==null) return null;
        if(map.containsKey(head)){
            return map.get(head);
        }
        map.put(head, new Node(head.val));
        for(Node node: head.neighbors){
            map.get(head).neighbors.add(dfs(node, map));
        }
        return map.get(head);
    }
}